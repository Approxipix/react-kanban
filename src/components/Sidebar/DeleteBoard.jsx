import React, { Component } from 'react';
import { connect } from 'react-redux';
import { history } from "../../redux/store";
import { bindActionCreators } from 'redux';
import { deleteBoard } from '../../redux/boardReducer/actions';
import { Text, SubmitButton } from '../BaseComponent';
import styled from 'styled-components'

const Container = styled.div`
  position: relative;
`;

class DeleteBoard extends Component {
  deleteBoard = () => {
    this.props.actions.deleteBoard({
      boardId: this.props.boardId
    });
    history.push('/boards')
  };

  render() {
    return (
      <Container>
        <Text>
          All actions will be removed from the activity feed and you wonâ€™t be able to re-open the board.
          There is no undo.
        </Text>
        <SubmitButton
          color={'danger'}
          onClick={() => this.deleteBoard()}
        >
          Delete board
        </SubmitButton>
      </Container>
    );
  }
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      deleteBoard: deleteBoard,
    }, dispatch)
  };
}

export default connect(null, mapDispatchToProps)(DeleteBoard);
