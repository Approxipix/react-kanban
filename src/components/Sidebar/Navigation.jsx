import React, { Component } from 'react';
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import styled from 'styled-components'

const Title = styled.h3`
  font-weight: 600;
`;

const List = styled.ul`
`;

const Item = styled.li`
  margin-bottom: .5rem;
  padding: .3rem .5rem;
  border-radius: .2rem;
  transition: background-color .2s ease-in;
  cursor: pointer;
  &:hover {
    background-color: rgba(9,30,66,.13);
  }
`;

const Link = styled.a`
  display: flex;
  align-items: center;
  font-size: .875rem;
  color: #172b4d;
  font-weight: 700;
`;

const Icon = styled.i`
  margin-right: .75rem;
  color: #42526e;
`;

class Navigation extends Component {
  constructor(props) {
    super(props);
    this.state = {
      component: null,
      editTitle: false,
    }
  }

  changeView = (e, item) => {
    e.preventDefault();
    this.props.changeView(item.component, item.title)
  };

  render() {
    const navigationItem = [
      {
        title: 'Change Background',
        component: 'ChangeBackground',
        icon: 'fill-drip',
      },
      // {
      //   title: '!= Search Cards',
      //   component: '',
      //   icon: '',
      // },
      // {
      //   title: '!= Copy Board',
      //   component: '',
      //   icon: '',
      // },
      {
        title: 'Delete Board',
        component: 'DeleteBoard',
        icon: 'trash',
      },
    ];

    return (
      <List>
        {navigationItem.map((item, index) => (
          <Item key={index} >
            <Link onClick={(e) => this.changeView(e, item)}>
              <Icon>
                <FontAwesomeIcon icon={item.icon} />
              </Icon>
              <Title>
                {item.title}
              </Title>
            </Link>
          </Item>
        ))}
      </List>
    );
  }
}

export default Navigation;
