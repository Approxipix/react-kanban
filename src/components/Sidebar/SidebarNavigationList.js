const SidebarNavigation = [
  {
    title: 'Change Background',
    component: 'BoardBackgroundEdit',
    icon: 'fill-drip',
  },
  // {
  //   title: '!= Search Cards',
  //   component: '',
  //   icon: '',
  // },
  // {
  //   title: '!= Copy Board',
  //   component: '',
  //   icon: '',
  // },
  {
    title: 'Delete Board',
    component: 'BoardDelete',
    icon: 'trash',
  },
];

export default SidebarNavigation
