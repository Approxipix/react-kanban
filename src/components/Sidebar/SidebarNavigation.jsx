import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import SidebarNavigationList from './SidebarNavigationList';
import styled from 'styled-components';

const Title = styled.h3`
  font-weight: 600;
`;

const List = styled.ul``;
List.displayName = 'List';

const Item = styled.li`
  margin-bottom: .5rem;
  padding: .3rem .5rem;
  border-radius: .2rem;
  transition: background-color .2s ease-in;
  cursor: pointer;
  &:hover {
    background-color: rgba(9,30,66,.13);
  }
`;

const Link = styled.a`
  display: flex;
  align-items: center;
  font-size: .875rem;
  font-weight: 700;
`;
Link.displayName = 'Link';

const Icon = styled.i`
  margin-right: .75rem;
  color: #42526e;
`;

class SidebarNavigation extends Component {
  constructor(props) {
    super(props);
    this.state = {
      navigation: SidebarNavigationList,
    }
  }

  changeView = (e, item) => {
    e.preventDefault();
    this.props.changeView(item.component, item.title)
  };

  render() {
    const { navigation } = this.state;
    return (
      <List>
        {navigation.map((item, index) => (
          <Item key={index} >
            <Link onClick={(e) => this.changeView(e, item)}>
              <Icon>
                <FontAwesomeIcon icon={item.icon} />
              </Icon>
              <Title>
                {item.title}
              </Title>
            </Link>
          </Item>
        ))}
      </List>
    );
  }
}

SidebarNavigation.defaultProps = {
  changeView: () => {},
};

SidebarNavigation.propTypes = {
  changeView: PropTypes.func.isRequired,
};

export default SidebarNavigation;
