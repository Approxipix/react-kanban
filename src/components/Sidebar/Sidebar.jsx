import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { toggleSidebar } from '../../redux/rootReducer/actions/actions';
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import SidebarNavigation from './SidebarNavigation';
import BoardBackgroundEdit from '../Board/BoardBackgroundEdit';
import BoardDelete from '../Board/BoardDelete';
import styled from 'styled-components';

const Wrapper = styled.div`
  position: absolute;
  top: 0;
  bottom: 0;
  width: 19rem;
  left: 100%;
  background-color: #F4F5F7;
  box-shadow: 0 .75rem 1.5rem -.375rem rgba(9,30,66,.25);
  transform: ${props => props.isOpened ? 'translateX(-100%)' : 'none'};
  transition: transform .3s ease-in-out;
`;

const Header = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: .75rem .5rem;
  border-bottom: 1px solid rgba(9,30,66,.13);
`;

const Title = styled.h3`
  font-size: 1rem;
  font-weight: 600;
`;

const Button = styled.button`
  color: #6b778c;
  font-size: 1rem;
  transition: color .1s ease-in;
  &:hover {
    color: #42526e;
  }
`;
Button.displayName = 'Button';

const Body = styled.div`
   padding: .75rem 1rem;
`;

class Sidebar extends Component {
  constructor(props) {
    super(props);
    this.state = {
      sidebarView: null,
      sidebarTitle: 'Menu',
    }
  }

  componentWillUnmount() {
    this.props.actions.toggleSidebar(false)
  }

  changeView = (component, title) => {
    this.setState({
      sidebarView: component,
      sidebarTitle: title,
    });
  };

  sidebarView = () => {
    const { boards, currentBoardID } = this.props;
    const { sidebarView } = this.state;

    let board;
    if (boards[currentBoardID]) {
      board = boards[currentBoardID]
    }

    switch (sidebarView) {
      case ('BoardBackgroundEdit'):
        return <BoardBackgroundEdit boardId={board._boardId} boardColor={board.color}/>;
      case ('BoardDelete'):
        return <BoardDelete boardId={board._boardId}/>;
      default:
        return <SidebarNavigation changeView={this.changeView}/>;
    }
  };

  render() {
    const { isSidebarOpened } = this.props;
    const { sidebarTitle, sidebarView } = this.state;
    return (
      <Wrapper isOpened={isSidebarOpened}>
        <Header>
          <Button
            onClick={() => this.changeView(null, 'Menu')}
            disabled={!sidebarView}
          >
            {!!sidebarView && <FontAwesomeIcon icon="chevron-left"/>}
          </Button>
          <Title>
            {sidebarTitle}
          </Title>
          <Button onClick={() => this.props.actions.toggleSidebar(false)}>
            <FontAwesomeIcon icon="times"/>
          </Button>
        </Header>
        <Body>
          {this.sidebarView()}
        </Body>
      </Wrapper>
    );
  }
}

Sidebar.defaultProps = {
  isSidebarOpened: false,
  boards: {},
};

Sidebar.propTypes = {
  currentBoardID: PropTypes.string,
  isSidebarOpened: PropTypes.bool,
  boards: PropTypes.objectOf(
    PropTypes.shape({
      _boardId: PropTypes.string,
      title: PropTypes.string,
      color: PropTypes.string.isRequired,
      lists: PropTypes.arrayOf(PropTypes.string)
    })
  ),
};


function mapStateToProps(state) {
  return {
    boards: state.boardReducer,
    currentBoardID: state.rootReducer.currentBoardID,
    isSidebarOpened: state.rootReducer.isSidebarOpened,
  }
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      toggleSidebar: toggleSidebar,
    }, dispatch)
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Sidebar);
