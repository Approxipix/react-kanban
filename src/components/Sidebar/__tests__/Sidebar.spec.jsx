import React from 'react';
import configureStore from 'redux-mock-store';
import { shallow } from 'enzyme';
import toJson from "enzyme-to-json";
import Sidebar from '../Sidebar';

const mockStore = configureStore([]);

describe('<Sidebar>', () => {
  let store;
  let component;

  beforeEach(() => {
    store = mockStore({
      rootReducer: {
        isSidebarOpened: false,
      }
    });
    store.dispatch = jest.fn();
    const state = {
      sidebarView: null,
      sidebarTitle: 'Menu',
    };

    component = shallow(<Sidebar store={store}  />).find('Sidebar').dive();
    component.setState({ ...state });
  });

  it('should render with given state from Redux store', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should dispatch an action on button click and after componentWillUnmount', () => {
    const expectedAction = {
      type: 'TOGGLE_SIDEBAR',
      payload: false
    };

    component.find('Button').last().simulate('click');

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(expectedAction);

    component.unmount();

    expect(store.dispatch).toHaveBeenCalledTimes(2);
  });

  it('checks if changeView sets sidebarView and sidebarTitle correctly', () => {
    component.instance().changeView('TestComponentView', 'Test Component');
    expect(component.instance().state.sidebarTitle).toEqual('Test Component');
    expect(component.instance().state.sidebarView).toEqual('TestComponentView');
  });

  it('it should render fontawesome icon if sidebarView was selected', () => {
    component.instance().changeView('TestComponentView', 'Test Component');

    expect(toJson(component)).toMatchSnapshot();
  });
});
