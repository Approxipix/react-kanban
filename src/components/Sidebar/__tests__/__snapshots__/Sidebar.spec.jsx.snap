// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sidebar> it should render fontawesome icon if sidebarView was selected 1`] = `
<styled.div
  isOpened={false}
>
  <styled.div>
    <Button
      disabled={false}
      onClick={[Function]}
    >
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon="chevron-left"
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        swapOpacity={false}
        symbol={false}
        title=""
        transform={null}
      />
    </Button>
    <styled.h3>
      Test Component
    </styled.h3>
    <Button
      onClick={[Function]}
    >
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon="times"
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        swapOpacity={false}
        symbol={false}
        title=""
        transform={null}
      />
    </Button>
  </styled.div>
  <styled.div>
    <SidebarNavigation
      changeView={[Function]}
    />
  </styled.div>
</styled.div>
`;

exports[`<Sidebar> should render with given state from Redux store 1`] = `
<styled.div
  isOpened={false}
>
  <styled.div>
    <Button
      disabled={true}
      onClick={[Function]}
    />
    <styled.h3>
      Menu
    </styled.h3>
    <Button
      onClick={[Function]}
    >
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon="times"
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        swapOpacity={false}
        symbol={false}
        title=""
        transform={null}
      />
    </Button>
  </styled.div>
  <styled.div>
    <SidebarNavigation
      changeView={[Function]}
    />
  </styled.div>
</styled.div>
`;
